### PR TITLE
Mazda FP 2.0 support

### PR DESCRIPTION
**Pull request by Jafaral**
ECU FW for CX-5 2017 and Mazda3 2017

Signed-off-by: Jafar Al-Gharaibeh <to.jafar@gmail.com>

### CODE REVIEW

**Comment by pd0wm** [comment]
Can you share a route with the fingerprint happening? Like to see how you ECUs respond to any of the existing queries we send.

Note you can also write this as
```python
MAZDA_VERSION_REQUEST = bytes([uds.SERVICE_TYPE.READ_DATA_BY_IDENTIFIER]) + \
  p16(uds.DATA_IDENTIFIER_TYPE.VEHICLE_MANUFACTURER_ECU_SOFTWARE_NUMBER)
```

**Comment by Jafaral** [comment]
@adeebshihadeh , any clue on why the unit tests are failing ? testing is a bit tricky for me since I'm stuck at 0.8.2 and can't run the master branch  directly on my car. I have to copy paste fixes from one branch to the other.

**Comment by Jafaral** [comment]
@jyoung8607 , Thanks. I did fix that. Now we have another failed test!

**Comment by jyoung8607** [comment]
I can repeat this intermittently on my machine. It's the Mock car tests, not the Mazda tests. Mock car startup is the worst case in terms of startup time, and with this PR, it looks like simulating 5 seconds of CAN traffic isn't *quite* long enough to consistently finish. Mock car startup requires falling through the VIN query, falling through all the UDS queries (and this PR adds a round of Mazda queries to try), falling through classic FPv1 with a7d4cf6, and then posting the onroad events the unittest wants to see.

Three varying examples:  [test_startup_log_fragment.txt](https://github.com/commaai/openpilot/files/6466664/test_startup_log_fragment.txt)

This seems sufficient to fix it for me:

```diff
diff --git a/selfdrive/controls/tests/test_startup.py b/selfdrive/controls/tests/test_startup.py
index ef5541b4..f85bc8b6 100755
--- a/selfdrive/controls/tests/test_startup.py
+++ b/selfdrive/controls/tests/test_startup.py
@@ -104,7 +104,7 @@ class TestStartup(unittest.TestCase):
     else:
       finger = _FINGERPRINTS[car_model][0]
 
-    for _ in range(500):
+    for _ in range(600):
       msgs = [[addr, 0, b'\x00'*length, 0] for addr, length in finger.items()]
       pm.send('can', can_list_to_can_capnp(msgs))
 
```

**Comment by Jafaral** [comment]
Thanks @jyoung8607 .  Should I incorporate this fix in this PR or a separate PR?

**Comment by pd0wm** [comment]
You can add it to this PR.

**Comment by jyoung8607** [comment]
Add it here for now, just to get you past this test and make sure there's no other broken CI tests hiding behind this one. After that, give @adeebshihadeh or @pd0wm time to review and comment. CI is their baby and they will decide if they like that change.

**Comment by jyoung8607** [comment]
While you're in [test_startup.py](selfdrive/controls/tests/test_startup.py), per adeeb's request you should probably go ahead and prepare for Mazda FPv1 deprecation. Add `CX5_FW_VERSIONS[]` in the same style as `COROLLA_FW_VERSIONS[]`, and reference it in the two Mazda tests the same way it's done in the Corolla test. Don't worry about the fuzzy FP stuff for now.

File: selfdrive/controls/tests/test_startup.py [hunk 1]
```diff
@@ -49,6 +56,9 @@ class TestStartup(unittest.TestCase):
     (EventName.startupMaster, HYUNDAI.KIA_STINGER, True, None),
     (EventName.communityFeatureDisallowed, HYUNDAI.KIA_STINGER, False, None),
 
+    # dashcamOnly, FW query
+    (EventName.startupNoControl, MAZDA.CX5, False, CX5_FW_VERSIONS),
+
```
Comment (jyoung8607) [inline selfdrive/controls/tests/test_startup.py:61]: Don't add a new test case, change the existing two.

File: selfdrive/controls/tests/test_startup.py [hunk 2]
```diff
@@ -49,6 +56,9 @@ class TestStartup(unittest.TestCase):
...
     # dashcamOnly car
     (EventName.startupNoControl, MAZDA.CX5, True, None),
     (EventName.startupNoControl, MAZDA.CX5, False, None),
```
Comment (jyoung8607) [inline selfdrive/controls/tests/test_startup.py:64]:
```suggestion
    (EventName.startupNoControl, MAZDA.CX5, True, CX5_FW_VERSIONS),
    (EventName.startupNoControl, MAZDA.CX5, False, CX5_FW_VERSIONS),
```

Comment (Jafaral) [reply selfdrive/controls/tests/test_startup.py:64]: I'm aware @jyoung8607 , was trying to get the new thing to work first. Thanks for your help addressing the failed tests.

**Comment by Jafaral** [comment]
response on  2017 CX-5 after some updates on my end. Waiting for more fingerprints from Mazda users. 
```
Found FW versions
{
  (Ecu.fwdRadar, 0x764, None): [b'K123-67XK2-F\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.eps, 0x730, None): [b'KJ01-3210X-G-00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.srs, 0x737, None): [b'KB7W-57KK2-A2\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.unknown, 0x726, None): [b'TK52-675X2-C-00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.unknown, 0x731, None): [b'GMB6-675S1-A-05\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.engine, 0x7e0, None): [b'PYNF-188K2-F\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.fwdCamera, 0x706, None): [b'B61L-67XK2-T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.hud, 0x7b2, None): [b'KB7W-55HK2-E\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.combinationMeter, 0x720, None): [b'K157-554K2-V\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.unknown, 0x761, None): [b'SH9N-189K2-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.transmission, 0x7e1, None): [b'PYNC-21PS1-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.unknown, 0x744, None): [b'B62W-675J1-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.esp, 0x760, None): [b'K123-437K2-E\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
  (Ecu.unknown, 0x756, None): [b'K123-430K2-C\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']
}
```

**Comment by Jafaral** [comment]
@pd0wm , I added a few modules  [here](https://github.com/Jafaral/openpilot/commit/cf4b96698c400906fc8b61a0473e9ca5ea2acc63#diff-d438db1e9b73b7398633d324df248932fced72703958b62c68e4d7688055c102R542)
Is this OK ? Do I need to update any other files? I still see these show as consts on the web interface,

**Comment by pd0wm** [comment]
Don't think we need all those ECUs. Every ECU we add also requires more upkeep of the database. Think the 5 ECUs you have right now are fine.

(we'd need to redeploy the web interface after adding the values, but as you've noticed capnp is very backwards-compatible so it just shows as integers if doesn't recognize the enum value.

**Comment by pd0wm** [comment]
Do you still expect more fingerprints? Would be best to confirm as soon as possible this request returns unique values. It's kind of a mess if we have to change this after merging.

**Comment by Jafaral** [comment]
@pd0wm , so far I collected 6 fingerprints, see  [here](https://github.com/Jafaral/openpilot/blob/mazda-0.8.4/selfdrive/car/mazda/values.py)

I collected all ECUs responses and I started trimming it down o get  a good unique subset. I'm down to 9 now, but as far as I can see, we maybe able to go to the "common five" and be OK. Or I can drop SSU and EPB  since I had to add definition for these and keep the remaining 7 which include:
eps
esp
engine
fwdRadar
fwdCamera
transmission
combinationMeter. 

I was hoping to collect more FPs, but users are slow to respond. If you are OK with the above first 5, or all 7 I can update the PR and move forward with the FP 2.0 support.

**Comment by pd0wm** [comment]
I'd prefer to keep only 5 if possible. If we don't get unique values we need to check if our request is getting enough data from the ECUs, but seems so far it's doing fine. Let's give people a few more days to respond, and see if we still don't see any duplicates.

**Comment by Jafaral** [comment]
Sure, the less the better. And yes, I didn't want to rush things before I see more data. We will see where we stand by the end of this week and go from there.

**Comment by Jafaral** [comment]
@pd0wm , I added one more car since with this FP:

```
    (Ecu.eps, 0x730, None): [
      b'TC3M-3210X-A-00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    (Ecu.engine, 0x7e0, None): [
      b'PXM4-188K2-C\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    (Ecu.fwdRadar, 0x764, None): [
      b'K131-67XK2-E\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    (Ecu.esp, 0x760, None): [
      b'TA0B-437K2-C\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', # 2020/2021
    ],
    (Ecu.fwdCamera, 0x706, None): [
      b'GSH7-67XK2-M\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    (Ecu.transmission, 0x7e1, None): [
      b'PXM4-21PS1-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    # IC - Instrument Cluster
    (Ecu.combinationMeter, 0x720, None): [
      b'TB4E-554K2-B\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    # SSU - Start Stop Unit
    (Ecu.ssu, 0x731, None): [
      b'K30W-675S1- -00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
    ],
    # Air Bag
    (Ecu.srs, 0x737, None): [
      b'TK48-57KK2-B41\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', # 2019/2021
    ],
    # Electronic parking brake
    (Ecu.epb, 0x756, None): [
      b'TA0A-430K2-A\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', # 2020/2021
    ],

```

I'm gong to drop the last 3 for sure. There is only one non-unique value in the remaining 7. I can drop transmission and combinationMeter as we talked before,  but I wonder if should keep those two and drop esp instead since it is shared between 2020 and 2021 models. 
Thoughts ?

I'll will get this PR updated based on your feedback

**Comment by Jafaral** [comment]
I went ahead and kept the original 5 plus transmission. Those seems to be a good combination for all cars. If you only want the original 5 , i can drop transmission.

**Comment by pd0wm** [comment]
Transmission is fine. This has been very unique on other brands.

**Comment by Jafaral** [comment]
Ready  for review @pd0wm

File: selfdrive/car/mazda/values.py [hunk 3]
```diff
@@ -72,6 +72,102 @@ class Buttons:
 }
 
 
+FW_VERSIONS = {
+  #
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/values.py:76]: remove these comments

File: selfdrive/car/fw_versions.py [hunk 4]
```diff
@@ -129,6 +134,13 @@ def p16(val):
     [VOLKSWAGEN_VERSION_RESPONSE],
     DEFAULT_RX_OFFSET,
   ),
+    # Mazda
```
Comment (adeebshihadeh) [inline selfdrive/car/fw_versions.py:137]: indentation

File: selfdrive/car/mazda/values.py [hunk 5]
```diff
@@ -72,6 +72,102 @@ class Buttons:
...
+      b'PXM7-21PS1-A\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
+    ],
+  },
+
+  #
+  # Mazda 3
+  #
+  CAR.Mazda3: {
+    (Ecu.eps, 0x730, None): [
+      b'K070-3210X-C-00\x00\x00\x00\x00\x00\x00\x00\x00\x00', # 2017
```
Comment (adeebshihadeh) [inline selfdrive/car/mazda/values.py:145]: remove these comments too

Comment (Jafaral) [reply selfdrive/car/mazda/values.py:145]: those at the end of the line?

Comment (adeebshihadeh) [reply selfdrive/car/mazda/values.py:145]: yes

**Comment by Jafaral** [comment]
pushed a commit addressing your comments @adeebshihadeh. I also dropped an unnecessary request in the FW query .

**Comment by pd0wm** [comment]
Looks good!

Can you remove all existing CAN fingerprints?

**Pull request merged**